File: mcp-server/observe_client.py
"""
Observation Sheet client for the MCP server.

Talks to the Google Apps Script endpoint that manages field observations
in the "Firefly Corner - Field Observations" Google Sheet.

Keeps separation of concerns: farmos_client.py talks to farmOS,
observe_client.py talks to the observation sheet.
"""

import json
import os
from typing import Optional

import requests
from dotenv import load_dotenv


class ObservationClient:
    """HTTP client for the Google Apps Script observation endpoint.

    The endpoint supports:
      GET  ?action=list[&status=...&section=...&observer=...&date=...&submission_id=...]
      POST {action: "update_status", updates: [{submission_id, status, reviewer, notes}]}
    """

    def __init__(self):
        self.endpoint = None
        self._connected = False

    @property
    def is_connected(self) -> bool:
        return self._connected

    def connect(self) -> bool:
        """Load endpoint URL from environment."""
        load_dotenv()

        self.endpoint = os.getenv("OBSERVE_ENDPOINT", "").rstrip("/")
        if not self.endpoint:
            raise ValueError(
                "Missing OBSERVE_ENDPOINT environment variable. "
                "Set it to the Google Apps Script deployment URL."
            )
        self._connected = True
        return True

    def list_observations(
        self,
        status: Optional[str] = None,
        section: Optional[str] = None,
        observer: Optional[str] = None,
        date: Optional[str] = None,
        submission_id: Optional[str] = None,
    ) -> dict:
        """Fetch observations from the Sheet with optional filters.

        Returns dict with keys: success, observations (list), count (int).
        """
        params = {"action": "list"}
        if status:
            params["status"] = status
        if section:
            params["section"] = section
        if observer:
            params["observer"] = observer
        if date:
            params["date"] = date
        if submission_id:
            params["submission_id"] = submission_id

        resp = requests.get(self.endpoint, params=params, timeout=30)
        resp.raise_for_status()
        return resp.json()

    def update_status(self, updates: list) -> dict:
        """Update review status of observation rows.

        Args:
            updates: List of dicts, each with:
                - submission_id (str): Target submission
                - status (str): New status (reviewed/approved/imported/rejected)
                - reviewer (str): Who is making the change
                - notes (str, optional): Review notes

        Returns dict with keys: success, updated (int), errors (list).
        """
        payload = json.dumps({
            "action": "update_status",
            "updates": updates,
        })
        # Content-Type: text/plain avoids CORS preflight (matching observe.js pattern).
        # Apps Script returns 302 redirect — requests follows as GET, which works.
        resp = requests.post(
            self.endpoint,
            data=payload,
            headers={"Content-Type": "text/plain"},
            timeout=30,
        )
        resp.raise_for_status()
        return resp.json()

File: mcp-server-ts/plugins/farm-plugin/src/clients/observe-client.ts
/**
 * Observation Sheet client — manages field observations via Apps Script.
 * Ported from Python observe_client.py, uses framework AxiosHttpClient.
 */

import { AppsScriptClient } from './apps-script-client.js';

export class ObservationClient extends AppsScriptClient {
  async listObservations(params: {
    status?: string;
    section?: string;
    observer?: string;
    date?: string;
    submission_id?: string;
  } = {}): Promise<any> {
    const q: Record<string, string> = { action: 'list' };
    if (params.status) q.status = params.status;
    if (params.section) q.section = params.section;
    if (params.observer) q.observer = params.observer;
    if (params.date) q.date = params.date;
    if (params.submission_id) q.submission_id = params.submission_id;
    return this.get(q);
  }

  async updateStatus(updates: Array<{
    submission_id: string;
    status: string;
    reviewer: string;
    notes?: string;
  }>): Promise<any> {
    return this.post({ action: 'update_status', updates });
  }

  async deleteImported(submissionId: string): Promise<any> {
    return this.post({ action: 'delete_imported', submission_id: submissionId });
  }

  async getMedia(submissionId: string): Promise<any> {
    return this.get({ action: 'get_media', submission_id: submissionId });
  }

  /**
   * Backfill helper — list Drive observation folders that have photos.
   * See Observations.gs handleListMediaFolders.
   */
  async listMediaFolders(date?: string): Promise<any> {
    const q: Record<string, string> = { action: 'list_media_folders' };
    if (date) q.date = date;
    return this.get(q);
  }

  /**
   * Backfill helper — fetch media files for a specific (date, section)
   * Drive folder directly, bypassing the Sheet submission lookup.
   */
  async getMediaByPath(date: string, section: string): Promise<any> {
    return this.get({ action: 'get_media_by_path', date, section });
  }
}
